src/main/java/com/limechain/runtime/hostapi/OffchainHostFunctions.java

```diff
@@ -213,4 +239,245 @@ public int extOffchainRandomSeed() {
+     * @param deadlinePointer a pointer-size to the SCALE encoded Option valud containing the UNIX timestamp.
+     * @return a pointer-size to the SCALE encoded Result value
+     */
+    public RuntimePointerSize extOffchainHttpResponseReadBodyVersion1(int requestId,
+                                                                      RuntimePointerSize bufferPointer,
+                                                                      RuntimePointerSize deadlinePointer) {
+        int timeout = timeoutFromDeadlineAndTimestamp(deadlinePointer, extOffchainTimestamp());
+        try {
+            HttpResponseType[] response = requests.getRequestsResponses(new int[]{requestId}, timeout);
+            if (response[0] == HttpResponseType.FINISHED) {
+                byte[] data = requests.getResponseBody(requestId);
+                hostApi.writeDataToMemory(data, bufferPointer);
+
+                byte[] resultValue = scaleEncodeIntResult(data.length);
+                return hostApi.writeDataToMemory(resultValue);
+            } else {
+                switch (response[0]){
+                    case INVALID_ID:
+                        return hostApi.writeDataToMemory(HttpResponseType.INVALID_ID.scaleEncodedResult());
+                    case DEADLINE_REACHED:
+                        return hostApi.writeDataToMemory(HttpResponseType.DEADLINE_REACHED.scaleEncodedResult());
+                    case IO_ERROR:
+                        return hostApi.writeDataToMemory(HttpResponseType.IO_ERROR.scaleEncodedResult());
+                    default:
+                        return null;
+                }
+            }
+        } catch (InvalidRequestId e) {
+            return hostApi.writeDataToMemory(HttpResponseType.INVALID_ID.scaleEncodedResult());
+        } catch (SocketTimeoutException e) {
+            log.log(Level.WARNING, e.getMessage(), e.getStackTrace());
+            return hostApi.writeDataToMemory(HttpResponseType.DEADLINE_REACHED.scaleEncodedResult());
+        } catch (IOException e) {
+            log.log(Level.WARNING, e.getMessage(), e.getStackTrace());
+            return hostApi.writeDataToMemory(HttpResponseType.IO_ERROR.scaleEncodedResult());
+        }
+    }
+
+    byte[] scaleEncodeHeaders(Map<String, List<String>> headers) throws IOException {
+        List<Pair<String, byte[]>> pairs = new ArrayList<>();
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        ScaleCodecWriter writer = new ScaleCodecWriter(outputStream);
+        for (String header : headers.keySet()) {
+            List<String> value = headers.get(header);
+            for (String headerValue : value) {
+                writer.writeByteArray(headerValue.getBytes());
+            }
+            pairs.add(new Pair<>(header, outputStream.toByteArray()));
+        }
+        return ScaleUtils.Encode.encode(pairs, String::getBytes, Function.identity());
```

Comment:
What I strongly suspect is, those multi-values of a single header key might need to be concatenated with an `" "` or `", "` as per HTTP specs, but I'm not entirely sure. Just raising this as a consideration.

In any case, a suggested nitpick to shorten the implementation (we can concat in a string, and then encode the string, I think we needn't the entire value-by-value encoding and then concatenation).
```suggestion
        List<Pair<String, String>> pairs = new ArrayList<>(headers.size());
        for (String header : headers.keySet()) {
            // NOTE: Not sure how the Spec expects us to handle repeated AND multi-valued headers, so we just concat the values (a little sus?)
            String value = String.join("", headers.get(header)); 
            pairs.add(new Pair<>(header, value));
        }
        return ScaleUtils.Encode.encode(pairs, String::getBytes, String::getBytes);
```